torch/optim/lars.py

```diff
@@ -0,0 +1,168 @@
+            params_with_grad = []
+            grads = []
+            momentum_buffer_list = []
+
+            for p in group["params"]:
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    grads.append(p.grad)
+
+                    state = self.state[p]
+
+                    momentum_buffer_list.append(state.get("momentum_buffer"))
+
+            lars(
+                params_with_grad,
+                grads,
+                momentum_buffer_list,
+                lr=group["lr"],
+                weight_decay=group["weight_decay"],
+                momentum=group["momentum"],
+                dampening=group["dampening"],
+                nesterov=group["nesterov"],
+                eps=self.eps,
+                trust_coefficient=self.trust_coefficient,
+                maximize=group["maximize"],
+            )
+
+            for p, momentum_buffer in zip(params_with_grad, momentum_buffer_list):
+                state = self.state[p]
+                state["momentum_buffer"] = momentum_buffer
+
+        return loss
+
+
+def lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    dampening: float,
+    nesterov: bool,
+    eps: float,
+    trust_coefficient: float,
+    maximize: bool,
+):
+    if not torch.jit.is_scripting():
+        func = _single_tensor_lars
+
+    func(
+        params,
+        grads,
+        momentum_buffer_list,
+        lr=lr,
+        weight_decay=weight_decay,
+        momentum=momentum,
+        dampening=dampening,
+        nesterov=nesterov,
+        eps=eps,
+        trust_coefficient=trust_coefficient,
+        maximize=maximize,
+    )
+
+
+def _single_tensor_lars(
+    params: List[Tensor],
+    grads: List[Tensor],
+    momentum_buffer_list: List[Optional[Tensor]],
+    *,
+    lr: float,
+    weight_decay: float,
+    momentum: float,
+    dampening: float,
+    nesterov: bool,
+    eps: float,
+    trust_coefficient: float,
+    maximize: bool,
+):
+    for i, param in enumerate(params):
+        d_p = grads[i] if not maximize else -grads[i]
+
+        p_norm = torch.norm(param.data)
+        g_norm = torch.norm(d_p.data)
+
+        if weight_decay != 0:
+            # LARS scaling:
+            if p_norm * g_norm > 0:
+                lars_lr = trust_coefficient * p_norm / (g_norm + p_norm * weight_decay + eps)
+
+                d_p = d_p.add(param, alpha=weight_decay)
+                d_p.mul_(lars_lr)
+
+        if momentum != 0:
+            buf = momentum_buffer_list[i]
+
+            if buf is None:
+                buf = torch.clone(d_p).detach()
+                momentum_buffer_list[i] = buf
+            else:
+                buf.mul_(momentum).add_(d_p, alpha=1 - dampening)
+
+            d_p = d_p.add(buf, alpha=momentum) if nesterov else buf
+
+    param.add_(d_p, alpha=-lr)
```

Comment:
the global LR is not applied to the momentum part to the best of my understanding

Reply:
I will check again the paper and some online implementations the next days.